File: src/net/minecraft/client/render/block/entity/PistonBlockEntityRenderer.cpp
#include "net/minecraft/client/render/block/entity/PistonBlockEntityRenderer.hpp"
#include "net/minecraft/block/Block.hpp"
#include "net/minecraft/block/PistonBlock.hpp"
#include "net/minecraft/block/PistonHeadBlock.hpp"
#include "net/minecraft/block/entity/PistonBlockEntity.hpp"
#include "net/minecraft/client/Minecraft.hpp"
#include "net/minecraft/client/gl/GlState.hpp"
#include "net/minecraft/client/render/Tessellator.hpp"
#include "net/minecraft/client/gl/Lighting.hpp"
namespace net::minecraft::client::render::block::entity {
void PistonBlockEntityRenderer::setWorld(net::minecraft::World* world) {
  blockRenderManager.setBlockView(world);
}
void PistonBlockEntityRenderer::render(
    const net::minecraft::block::entity::BlockEntity& blockEntity, double x, double y, double z, float tickDelta) {
  const auto* piston = dynamic_cast<const net::minecraft::block::entity::PistonBlockEntity*>(&blockEntity);
  if(piston == nullptr) {
    return;
  }
  net::minecraft::block::Block* block =
      net::minecraft::block::Block::BLOCKS[static_cast<std::size_t>(piston->getPushedBlockId())];
  if(block == nullptr || piston->getProgress(tickDelta) >= 1.0f) {
    return;
  }
  Tessellator& tessellator = render::INSTANCE;
  bindTexture("/terrain.png");
  gl::Lighting::turnOff();
  const gl::preset::PistonTranslucent pistonCaps;
  tessellator.startQuads();
  tessellator.translate(static_cast<double>(static_cast<float>(x) - static_cast<float>(piston->x) +
                                            piston->getRenderOffsetX(tickDelta)),
                        static_cast<double>(static_cast<float>(y) - static_cast<float>(piston->y) +
                                            piston->getRenderOffsetY(tickDelta)),
                        static_cast<double>(static_cast<float>(z) - static_cast<float>(piston->z) +
                                            piston->getRenderOffsetZ(tickDelta)));
  tessellator.color(1, 1, 1);
  if(block == net::minecraft::block::Block::PISTON_HEAD && piston->getProgress(tickDelta) < 0.5f) {
    blockRenderManager.renderPistonHeadWithoutCulling(block->id, piston->x, piston->y, piston->z, false);
  } else if(piston->isSource() && !piston->isExtending()) {
    auto* pistonBlock = dynamic_cast<net::minecraft::block::PistonBlock*>(block);
    if(pistonBlock != nullptr && net::minecraft::block::Block::PISTON_HEAD != nullptr) {
      const int blockMeta = blockRenderManager.ctx.blockView->getBlockMeta(piston->x, piston->y, piston->z);
      const int facing = net::minecraft::block::PistonHeadBlock::getFacing(blockMeta);
      blockRenderManager.ctx.faceTextureOverride = pistonBlock->getTopTexture();
      blockRenderManager.ctx.faceTextureSide = facing;
      blockRenderManager.renderPistonHeadWithoutCulling(net::minecraft::block::Block::PISTON_HEAD->id,
                                                        piston->x,
                                                        piston->y,
                                                        piston->z,
                                                        piston->getProgress(tickDelta) < 0.5f);
      blockRenderManager.ctx.faceTextureOverride = -1;
      blockRenderManager.ctx.faceTextureSide = -1;
      tessellator.translate(static_cast<double>(static_cast<float>(x) - static_cast<float>(piston->x)),
                            static_cast<double>(static_cast<float>(y) - static_cast<float>(piston->y)),
                            static_cast<double>(static_cast<float>(z) - static_cast<float>(piston->z)));
      blockRenderManager.renderExtendedPiston(block->id, piston->x, piston->y, piston->z);
    }
  } else {
    blockRenderManager.renderWithoutCulling(block->id, piston->x, piston->y, piston->z);
  }
  tessellator.translate(0.0, 0.0, 0.0);
  tessellator.draw();
  gl::Lighting::turnOn();
}
} // namespace net::minecraft::client::render::block::entity

File: mcp/src/net/minecraft/client/render/block/entity/PistonBlockEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.block.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.block.Block;
import net.minecraft.block.PistonBlock;
import net.minecraft.block.entity.PistonBlockEntity;
import net.minecraft.client.Minecraft;
import net.minecraft.client.render.Tessellator;
import net.minecraft.client.render.block.BlockRenderManager;
import net.minecraft.client.render.block.entity.BlockEntityRenderer;
import net.minecraft.client.render.platform.Lighting;
import net.minecraft.world.World;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class PistonBlockEntityRenderer
extends BlockEntityRenderer {
    private BlockRenderManager blockRenderManager;

    public void render(PistonBlockEntity arg, double d, double e, double f, float g) {
        Block block = Block.BLOCKS[arg.getPushedBlockId()];
        if (block != null && arg.getProgress(g) < 1.0f) {
            Tessellator tessellator = Tessellator.INSTANCE;
            this.bindTexture("/terrain.png");
            Lighting.turnOff();
            GL11.glBlendFunc(770, 771);
            GL11.glEnable(3042);
            GL11.glDisable(2884);
            if (Minecraft.isAmbientOcclusionEnabled()) {
                GL11.glShadeModel(7425);
            } else {
                GL11.glShadeModel(7424);
            }
            tessellator.startQuads();
            tessellator.translate((double)((float)d - (float)arg.x + arg.getRenderOffsetX(g)), (double)((float)e - (float)arg.y + arg.getRenderOffsetY(g)), (double)((float)f - (float)arg.z + arg.getRenderOffsetZ(g)));
            tessellator.color(1, 1, 1);
            if (block == Block.PISTON_HEAD && arg.getProgress(g) < 0.5f) {
                this.blockRenderManager.renderPistonHeadWithoutCulling(block, arg.x, arg.y, arg.z, false);
            } else if (arg.isSource() && !arg.isExtending()) {
                Block.PISTON_HEAD.setSprite(((PistonBlock)block).getTopTexture());
                this.blockRenderManager.renderPistonHeadWithoutCulling(Block.PISTON_HEAD, arg.x, arg.y, arg.z, arg.getProgress(g) < 0.5f);
                Block.PISTON_HEAD.clearSprite();
                tessellator.translate((double)((float)d - (float)arg.x), (double)((float)e - (float)arg.y), (double)((float)f - (float)arg.z));
                this.blockRenderManager.renderExtendedPiston(block, arg.x, arg.y, arg.z);
            } else {
                this.blockRenderManager.renderWithoutCulling(block, arg.x, arg.y, arg.z);
            }
            tessellator.translate(0.0, 0.0, 0.0);
            tessellator.draw();
            Lighting.turnOn();
        }
    }

    public void setWorld(World world) {
        this.blockRenderManager = new BlockRenderManager(world);
    }
}

